app/services/journal_row_builder.rb

```diff
@@ -0,0 +1,165 @@
+    @recharge_enabled = SettingsHelper.feature_on?(:recharge_accounts)
+  end
+
+  # Idempotent method that builds an array of new journal_rows. Also adds to
+  # errors if any problems occur. Does not save anything to the database.
+  # Returns self to support chaining.
+  def build
+    reset
+    order_details.each do |order_detail|
+      validate_order_detail(order_detail)
+      new_journal_rows = order_detail_to_journal_rows(order_detail)
+      update_product_recharges(order_detail)
+      @journal_rows.concat(new_journal_rows)
+    end
+    add_journal_rows_from_product_recharges
+    self
+  end
+
+  # Creates journal rows for the given order_details.
+  # Does not gracefully handle unexpected database failures.
+  # Wrapping everything in a transaction was not an option because oracle will
+  # only support 1000 transactions per request.
+  # Returns self to support chaining.
+  def create
+    build
+    if valid? && journal_rows.present?
+      journal_rows.each { |journal_row| JournalRow.create!(journal_row) }
+      set_journal_for_order_details(journal, order_details.map(&:id))
+    end
+    self
+  end
+
+  # Returns true if errors are not present; otherwise false.
+  # Best to use this incase errors implementation changes.
+  def valid?
+    errors.blank?
+  end
+
+  # Set an array of facility_ids with pending journals
+  def pending_facility_ids
+    @pending_facility_ids ||= Journal.facility_ids_with_pending_journals
+  end
+
+  # Run all validations on an order detail
+  def validate_order_detail(order_detail)
+    validate_order_detail_unjournaled(order_detail)
+    validate_facility_can_journal(order_detail)
+    validate_account(order_detail)
+  end
+
+  # Validate the order_detail hasn't already been journaled
+  def validate_order_detail_unjournaled(order_detail)
+    if order_detail.journal_id.present?
+      @errors << "#{order_detail} is already journaled in journal #{order_detail.journal_id}"
+    end
+  end
+
+  # Validate the order_details's facility doesn't currently have a pending
+  # journal.
+  #
+  # Ported from original journal model logic. Consider replacing the raised
+  # error with a push to `@errors` instead.
+  def validate_facility_can_journal(order_detail)
+    facility_id = order_detail.order.facility_id
+    unless journaled_facility_ids.member?(facility_id)
+      if pending_facility_ids.member?(facility_id)
+        raise ::Journal::CreationError.new(I18n.t(
+          "activerecord.errors.models.journal.pending_overlap",
+          label: order_detail.to_s,
+          facility: Facility.find(facility_id)
+        ))
+      else
+        journaled_facility_ids.add(facility_id)
+      end
+    end
+  end
+
+  # Validate each journal_row account. This is necessary because an order_detail
+  # may generate multiple journal_rows with different accounts, and we need to
+  # validate them all.
+  #
+  # TODO: we may need to add a validator factory for split_accounts. Otherwise
```

Comment:
I believe it is a safe assumption that the product's account cannot be a split account.